### PR TITLE
Fix deliveryman

### PR DESCRIPTION
**Pull request by Raycoms**
# Changes proposed in this pull request:
 Improves the dman:
- Don't gather after deliver
- Don't deliver the same directly
- Don't gather requested items.

Review please

### CODE REVIEW

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 63 issues
* ![CRITICAL][CRITICAL] 6 critical
* ![MAJOR][MAJOR] 56 major
* ![MINOR][MINOR] 1 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L230](https://github.com/Minecolonies/minecolonies/blob/1b6afaecd317cc0ea5061abd3f59473851d157e2/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L230): Remove this call from a constructor to the overridable "setupRsDataStore" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1699)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1355](https://github.com/Minecolonies/minecolonies/blob/1b6afaecd317cc0ea5061abd3f59473851d157e2/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1355): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1425](https://github.com/Minecolonies/minecolonies/blob/1b6afaecd317cc0ea5061abd3f59473851d157e2/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1425): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1456](https://github.com/Minecolonies/minecolonies/blob/1b6afaecd317cc0ea5061abd3f59473851d157e2/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1456): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1468](https://github.com/Minecolonies/minecolonies/blob/1b6afaecd317cc0ea5061abd3f59473851d157e2/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1468): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1571](https://github.com/Minecolonies/minecolonies/blob/1b6afaecd317cc0ea5061abd3f59473851d157e2/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1571): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/1b6afaecd317cc0ea5061abd3f59473851d157e2/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/1b6afaecd317cc0ea5061abd3f59473851d157e2/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/1b6afaecd317cc0ea5061abd3f59473851d157e2/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/1b6afaecd317cc0ea5061abd3f59473851d157e2/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java [hunk 1]
```diff
@@ -1269,7 +1269,15 @@ public boolean buildingRequiresCertainAmountOfItem(final ItemStack stack, final
      */
     public Map<Predicate<ItemStack>, Integer> getRequiredItemsAndAmount()
     {
-        return keepX;
+        final Map<Predicate<ItemStack>, Integer> toKeep = new HashMap<>();
+        toKeep.putAll(keepX);
+        final IRequestManager manager = colony.getRequestManager();
+        toKeep.put(stack -> this.getOpenRequestsByCitizen().values().stream()
+                .anyMatch(list -> list.stream()
+                        .anyMatch(token -> manager.getRequestForToken(token) instanceof IDeliverable
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java]: You are missing it here. You need the request.

**Review by marchermans — CHANGES_REQUESTED** [review]

**Review by marchermans — APPROVED** [review]

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 63 issues
* ![CRITICAL][CRITICAL] 6 critical
* ![MAJOR][MAJOR] 56 major
* ![MINOR][MINOR] 1 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L230](https://github.com/Minecolonies/minecolonies/blob/606c7a3eecb5df5bdc22e43cbb6d68274ac48405/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L230): Remove this call from a constructor to the overridable "setupRsDataStore" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1699)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1355](https://github.com/Minecolonies/minecolonies/blob/606c7a3eecb5df5bdc22e43cbb6d68274ac48405/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1355): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1425](https://github.com/Minecolonies/minecolonies/blob/606c7a3eecb5df5bdc22e43cbb6d68274ac48405/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1425): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1456](https://github.com/Minecolonies/minecolonies/blob/606c7a3eecb5df5bdc22e43cbb6d68274ac48405/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1456): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1468](https://github.com/Minecolonies/minecolonies/blob/606c7a3eecb5df5bdc22e43cbb6d68274ac48405/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1468): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1571](https://github.com/Minecolonies/minecolonies/blob/606c7a3eecb5df5bdc22e43cbb6d68274ac48405/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1571): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/606c7a3eecb5df5bdc22e43cbb6d68274ac48405/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/606c7a3eecb5df5bdc22e43cbb6d68274ac48405/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/606c7a3eecb5df5bdc22e43cbb6d68274ac48405/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/606c7a3eecb5df5bdc22e43cbb6d68274ac48405/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by lunadragnstar** [comment]
It would also be nice if the deliveryman didn't drop items in chests instead of warehouse. (I found the lumberjack chest full of eggs from the chickenherder. They should be dropping that off at the warehouse.

**Pull request merged**